common/ffitools.go

```diff
@@ -0,0 +1,28 @@
+package common
+
+/*
+	#include <stdint.h>
+
+	/// free the memory allocated and returned by extern functions by transferring ownership back to
+	/// Rust. must be called on each pointer returned by the functions precisely once to ensure safety.
+	extern void freeptr(uint8_t *ptr);
+*/
+import "C"
+import (
+	"fmt"
+)
+
+// Common errors used in various packages
```

Comment:
These are all used in multiple places. It seems silly to define the same error or util in multiple places, esp. when it's being used the same way (as part of the FFI bridge).

Reply:
- `KeyLenErr` is only used in a single place (and in that specific case will never be returned by `ReadPubkeyFromLedger`
- `PathErr` is also only used in a single place (in `bip32`)
- `PointerErr` is just "some error happened in non-go code" and can be made more specific by the Rust code returning error codes

I don't doubt that some errors might overlap between `remote-wallet` and `bip32`, but one is a wrapper of `lspacemesh_remote_wallet` and one of `led25519_bip32`. Errors are defined by those dependencies; do they share a `common` package defining them?

As an alternative I think having just one package "smkeys" with both `DeriveBip32` and `PubKeyFromLedger` definitions in it can also make sense. Technically they are different, but functionally they do the same things - deriving a key from a derivation path.

As a second alternative I also think would work nicely is the following package layout:

- `smkeys`: contains shared types like errors and interfaces that are implemented by types in the other packages
- `smkeys/bip32`: implements the interfaces in `smkeys` on a software level using `bip32`
- `smkeys/ledger`: implements the interfaces in `smkeys` for a ledger wallet

Similar to how e.g. `crypto` is an abstraction with interfaces and types and `crypto/sha256`, `crypto/md5`, etc. implement those abstractions.

Reply:
ended up removing this module completely